### ISSUE TITLE
Add support for multi-dimensional ARRAY types

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#252]
H2 supports this (dimensions may differ, just like in Java):

```sql
select array[array[], array[1], array[1, 2]]
```

PostgreSQL supports this (all dimensions must be equal):

```sql
select array[array[1, 2], array[2, 3], array[3, 4]]
```

jOOQ should formally support these:

- [x] In the code generator
  - [x] Java
  - [x] Scala
  - [x] Kotlin
  - [x] Dialects
    - [x] H2
    - [x] ~HSQLDB~ (postponed, see https://github.com/jOOQ/jOOQ/issues/13987)
    - [x] PostgreSQL
  - [x] Generate `Arrays.deepToString()` calls instead of `Arrays.toString()`
- [x] In the runtime
  - [x] As DDL
  - [x] In casts
  - [x] As bind values / inline values
    - [x] inline values are currently serialised incorrectly, e.g. `'{"{\"a\",\"b\"}","{\"c\",\"d\"}"}'`
  - [x] In different operators, such as `ARRAY_AGG()` or `ARRAY()`
  - [x] `AbstractStore::equals` has to use `Arrays::deepEquals` instead of `Arrays::equals`. `hashCode()` doesn't need to be adapted, because we currently ignore array hash codes in the `Record` hash code calculation.
- [x] In the parser

**Comment by lukaseder** [comment]
#13454 has been fixed and integration tests confirm that `ARRAY_AGG(array_col)` work on at least H2 and PostgreSQL (HSQLDB, CockroachDB don't support nested arrays).

I might look a bit more into this topic soon, given that the feature request has received 10 upvotes already by now.

**Comment by lukaseder** [comment]
So, working on this now. Confirmed, `ARRAY_AGG(...)` or `ARRAY[ARRAY[...]]` already work out of the box, and so do bind variables, as jOOQ just passes them through to JDBC drivers.

I guess this might be mostly about code generation support, then.

**Comment by lukaseder** [comment]
On H2, this fun query helps look up array types recursively:

```sql
with recursive 
  e (object_catalog, object_schema, object_name, root_collection_type_identifier, collection_type_identifier, data_type, character_maximum_length, numeric_precision, numeric_scale, datetime_precision, dtd_identifier, suffix) as (
    select object_catalog, object_schema, object_name, collection_type_identifier, collection_type_identifier, data_type, character_maximum_length, numeric_precision, numeric_scale, datetime_precision, dtd_identifier, ' ARRAY' suffix
    from information_schema.element_types
    union all
    select e2.object_catalog, e2.object_schema, e2.object_name, e1. root_collection_type_identifier, e2.collection_type_identifier, e2.data_type, e2.character_maximum_length, e2.numeric_precision, e2.numeric_scale, e2.datetime_precision, e2.dtd_identifier, e1.suffix || ' ARRAY' suffix
    from e e1
    join information_schema.element_types e2
      on e1.object_catalog = e2.object_catalog
      and e1.object_schema = e2.object_schema
      and e1.object_name = e2.object_name
      and e1.dtd_identifier = e2.collection_type_identifier
  )
select c.column_name, c.data_type, e.data_type || e.suffix, *
from information_schema.columns c
left join e 
  on c.dtd_identifier = e.root_collection_type_identifier 
  and e.data_type != 'ARRAY'
where c.table_name = 'T_ARRAYS'
order by c.ordinal_position;
```

Resulting in:

```
|COLUMN_NAME             |DATA_TYPE|E.DATA_TYPE || E.SUFFIX            |
|------------------------|---------|-----------------------------------|
|ID                      |INTEGER  |                                   |
|STRING_ARRAY            |ARRAY    |CHARACTER VARYING ARRAY            |
|NUMBER_ARRAY            |ARRAY    |INTEGER ARRAY                      |
|DATE_ARRAY              |ARRAY    |DATE ARRAY                         |
|STRING_ARRAY_ARRAY      |ARRAY    |CHARACTER VARYING ARRAY ARRAY      |
|STRING_ARRAY_ARRAY_ARRAY|ARRAY    |CHARACTER VARYING ARRAY ARRAY ARRAY|
|NUMBER_ARRAY_ARRAY      |ARRAY    |INTEGER ARRAY ARRAY                |
|NUMBER_ARRAY_ARRAY_ARRAY|ARRAY    |INTEGER ARRAY ARRAY ARRAY          |
```

It's good to keep this in mind for the future. At first, I don't want to rely on recursion in H2. It's been too buggy in the past. Alternative options include:

- Manually left joining `ELEMENT_TYPES` e.g. 8x
- Pulling the entire `ELEMENT_TYPES` table into memory and doing the recursion there, only once

The latter option is probably the best choice here

**Comment by lukaseder** [comment]
In PostgreSQL, things are a bit more tricky as we're not using any formal SQL standards compliant type declaration, such as in H2 (`INTEGER ARRAY ARRAY`). In fact, we're not fetching the dimensions at all, which are available only through `pg_attribute.attndims`, not in the type name's `_` prefix. See also: https://github.com/jOOQ/jOOQ/issues/13970

I might fix https://github.com/jOOQ/jOOQ/issues/13970 later after implementing this feature here. The code generator so far assumes that `int` is the base type, `_int` is `int[]`, `__int` is `int[][]`, but this isn't the case.

**Comment by lukaseder** [comment]
Well, how about that. In PostgreSQL, multi dimensional arrays are just a hint, not something that is actually enforced:

```sql
drop table t;
create table t (i0 int, i1 int[], i2 int[][], i3 int[][][]);
insert into t values (1, array[array[array[1]]], array[array[1]], array[1]);
select * from t;
```

Result:

```
|i0 |i1     |i2   |i3 |
|---|-------|-----|---|
|1  |{{{1}}}|{{1}}|{1}|
```

A thing is either an array of a base type (and in case of tables, the dimensions are stored, for documentation purposes?), or a scalar type. In fact, functions don't even retain their dimension specs:

```sql
create function f (p0 int, p1 int[], p2 int[][], p3 int[][][])
returns int[][][][]
as $$
begin
  return array[p3];
end;
$$ language plpgsql;

select pg_get_functiondef(p.oid)
from pg_proc as p
join pg_type as t on p.prorettype = t.oid
where p.proname = 'f';
```

Produces:

```sql
CREATE OR REPLACE FUNCTION public.f(p0 integer, p1 integer[], p2 integer[], p3 integer[])
 RETURNS integer[]
 LANGUAGE plpgsql
AS $function$
begin
  return array[p3];
end;
$function$
```

This is underwhelming :-/ Hence, the code generator can't really produce the correct dimensions on function API (unless users enforce it with forced types)...

**Comment by lukaseder** [comment]
A first draft has been pushed. Further work include:

- Refactorings to re-use query fragments
- UDT support for multi dimensional arrays (also testing multi dimensional UDT arrays containing other multi dimensional UDT arrays, returned from table valued functions!)
- Regression tests on CockroachDB, YugabyteDB

**Comment by katzyn** [comment]
In H2 2.\*.\* you can use
```SQL
CREATE TABLE TEST(ID INT, V INTEGER ARRAY ARRAY[10] ARRAY);

SELECT COLUMN_NAME, DATA_TYPE_SQL('PUBLIC', 'TEST', 'TABLE', DTD_IDENTIFIER)
    FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_NAME = 'TEST'
    ORDER BY ORDINAL_POSITION;
> ID	INTEGER
> V	INTEGER ARRAY ARRAY[10] ARRAY
```
instead of recursive query, but [`DATA_TYPE_SQL`](https://h2database.com/html/functions.html#data_type_sql) is not configurable and it always returns a complete normalized data type with all its parameters, so it can return something like `ROW("A" NUMERIC(5), "B" ENUM('a', 'b') ARRAY[20], "C" INTERVAL DAY(10) TO SECOND(9)) ARRAY[100]` for array column.

It is possible to include a test case with your recursive query into test cases of H2 to avoid regressions with it in future versions of H2.

**Comment by lukaseder** [comment]
Thanks for the heads up.

> `ARRAY[10]`

Does that describe a max length of an array?



> instead of recursive query, but [`DATA_TYPE_SQL`](https://h2database.com/html/functions.html#data_type_sql)

The solution I've found is acceptable. I'm just loading all of `ELEMENT_TYPES` into memory and recurse over the data structure in Java. Just because I can use SQL doesn't mean I have to use it for *everything* 😅.

I'm wary of using such functions in the code generator as it's hard to be sure:

- If it has always existed in every RDBMS version that jOOQ supports
- If it will continue to exist in this way

Specifically, H2 has a track record of changing such things, and I wouldn't want to run this risk. `ELEMENT_TYPES` is a standard `INFORMATION_SCHEMA` view, so I think it will be more reliable for jOOQ's code generation needs.

> It is possible to include a test case with your recursive query into test cases of H2 to avoid regressions with it in future versions of H2.

That's certainly an option, but there's no guarantee that these queries remain the same. Also, jOOQ still supports H2 1.4, so I don't want to write 2-3 queries, one for each different H2 version...

**Comment by katzyn** [comment]
> Does that describe a max length of an array?

Yes, in the SQL Standard and H2 arrays may have optional maximum cardinality, you can also read it from `MAXIMUM_CARDINALITY` column in `INFORMATION_SCHEMA.COLUMNS`, `DOMAINS`, `ELEMENT_TYPES`, `FIELD`, `ROUTINES`, `PARAMETERS`, and non-standard `CONSTANTS` table.

**Comment by lukaseder** [comment]
Thanks for the explanations! jOOQ is never finished, I suspect. Here we go: https://github.com/jOOQ/jOOQ/issues/13972

**Comment by lukaseder** [comment]
Implemented for jOOQ 3.18.0

**Comment by lukaseder** [comment]
That last example likely doesn't work in Scala/Kotlin yet: https://github.com/jOOQ/jOOQ/issues/13980. I've planned some more robust integration tests for those generators soon, so I'll fix this when that's ready.

**Comment by lukaseder** [comment]
For HSQLDB support, see https://github.com/jOOQ/jOOQ/issues/13987
